spec/ddtrace/sampler_spec.rb

```diff
@@ -126,6 +126,34 @@
   end
 end
 
+RSpec.describe Datadog::RateByKeySampler do
+  subject(:sampler) { described_class.new(default_key, default_rate, &resolver) }
+
+  let(:default_key) { 'default-key' }
+
+  let(:span) { Datadog::Span.new(tracer, 'test-span') }
+  let(:resolver) { ->(span) { span.name } } # Resolve +span.name+ to the lookup key.
+
+  describe '#sample!' do
+    subject(:sample!) { sampler.sample!(span) }
+
+    # For testing purposes, never keep a span by default.
+    # DEV: Setting this to 0 would trigger a safe guard in `RateSampler` and set it to 100% instead.
```

Comment:
Would it be possible to turn this claim into another test case?

Reply:
This is tested at the `RateSampler`: https://github.com/DataDog/dd-trace-rb/blob/e391d2eb64d3c6151a4bdd2710c6a8c7c1d57457/spec/ddtrace/sampler_spec.rb#L59-L63

Do you think it's enough coverage? I always struggle with "wrapper" classes, and they normally requiring testing everything again... Specially complicated for non-trivial classes like these sampling ones, given some of the important behaviours are stochastic.

Reply:
There's something up with the validation, the tests still pass if I use 0:

```diff
diff --git a/spec/ddtrace/sampler_spec.rb b/spec/ddtrace/sampler_spec.rb
index 7a679018..92ed3924 100644
--- a/spec/ddtrace/sampler_spec.rb
+++ b/spec/ddtrace/sampler_spec.rb
@@ -139,7 +139,7 @@ RSpec.describe Datadog::RateByKeySampler do

     # For testing purposes, never keep a span by default.
     # DEV: Setting this to 0 would trigger a safe guard in `RateSampler` and set it to 100% instead.
-    let(:default_rate) { Float::MIN }
+    let(:default_rate) { 0 }
     it { is_expected.to be(false) }
```

spec output:

```bash
$ bundle exec rspec spec/ddtrace/sampler_spec.rb --format=progress
Run options: include {:focus=>true}

All examples were filtered out; ignoring {:focus=>true}

Randomized with seed 45767
......................................E, [2021-08-10T09:21:02.710208 #14099] ERROR -- ddtrace: [ddtrace] (/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/sampler.rb:52:in `initialize') sample rate is not between 0 and 1, disabling the sampler
.E, [2021-08-10T09:21:02.710834 #14099] ERROR -- ddtrace: [ddtrace] (/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/sampler.rb:52:in `initialize') sample rate is not between 0 and 1, disabling the sampler
......................

Finished in 0.07731 seconds (files took 0.73873 seconds to load)
61 examples, 0 failures

Randomized with seed 45767
```

> Do you think it's enough coverage? I always struggle with "wrapper" classes, as they technically require testing everything that was wrapped again... Specially complicated for non-trivial classes like these sampling ones, given some of the important behaviours are stochastic and verbose if duplicated.

I was going to say "I think so", until I tried putting in a 0 and it worked. I think the problem is that actually `Sampler` validates the rate in the constructor, but then allows `sample_rate=` to set whatever it wants, so the gap is actually in the `Sampler`.

But yes, in general this is all a balancing act in terms of value and effort, but I think in most cases it makes sense to write tests in a white-box manner, with some knowledge of what's inside, to avoid these kinds of repetitions, and then in a case-by-case basis decide when we should switch to a black-box approach.